zephyr/lib/regions_mm.c

```diff
@@ -1,53 +1,454 @@
+
+			if (allocators_bitaray == NULL) {
+				rfree(new_heap);
+				rfree(new_allocator);
+				return NULL;
+			}
+
+			size_t bitaray_size = SOF_DIV_ROUND_UP(SOF_DIV_ROUND_UP(
+				cfg->block_bundles_table[i].number_of_blocks, 8),
+					sizeof(uint32_t));
+
+			uint32_t *bitaray_bitfield;
+
+			bitaray_bitfield = rzalloc(SOF_MEM_ZONE_SYS, 0, SOF_MEM_CAPS_RAM,
+				sizeof(uint32_t) * bitaray_size);
+
+			if (bitaray_bitfield == NULL) {
+				rfree(new_heap);
+				rfree(new_allocator);
+				rfree(allocators_bitaray);
+				return NULL;
+			}
+
+			allocators_bitaray->num_bits =
+				cfg->block_bundles_table[i].number_of_blocks;
+			allocators_bitaray->num_bundles = bitaray_size;
+			allocators_bitaray->bundles = bitaray_bitfield;
+
+			new_allocator->num_blocks =
+				cfg->block_bundles_table[i].number_of_blocks;
+			new_allocator->blk_sz_shift =
+				ilog2(cfg->block_bundles_table[i].block_size);
+			new_allocator->buffer =
+				(uint8_t *)new_heap->virtual_region->addr +
+					offset;
+			new_allocator->bitmap = allocators_bitaray;
+
+			offset += cfg->block_bundles_table[i].number_of_blocks *
+				cfg->block_bundles_table[i].block_size;
+			new_heap->physical_blocks_allocators[i] = new_allocator;
+		}
+	}
+
+	new_heap->span = span;
+
+	/* Save the new heap pointer to a linked list */
+	list_item_append(&vm_heaps, &new_heap->node);
+
+	return new_heap;
+}
+
+/**
+ * @brief Alloc function
+ *
+ * Allocates memory on heap from provided heap pointer.
+ * Check if we need to map physical memory for given allocation
+ * and do it when nesscesary.
+ * @param heap pointer to a heap that will be used for allocation.
+ * @param alloc_size Size in bytes to allocate.
+ *
+ * @retval ptr to a new heap if successful
+ * @retval NULL on allocation failure
+ */
+void *vmh_alloc(struct virtual_memory_heap *heap, size_t alloc_size)
+{
+	void *ptr = NULL;
+	/* Error code we will ghater when allocating on physical block
+	 * allocators.
+	 * Loop will continiue as long as there are blocks to iterate on and
+	 * if allocation keeps returning error code.
+	 * When alloc succeeded it will return 0 on allocation and we break
+	 * form loop and continiue.
+	 */
+	bool allocation_error_code = true;
+	struct active_allocation *new_allocation =
+		rzalloc(SOF_MEM_ZONE_SYS, 0, SOF_MEM_CAPS_RAM,
+			sizeof(struct active_allocation));
+
+	if (new_allocation == NULL)
+		return NULL;
+
+	/* If span is on for heap allocation will be made on first available block
+	 * series on given block bundle.
+	 * For span off allocation will only happened on first free block
+	 * on first allocator that can accommodate requested size.
+	 */
+	for (size_t i = 0; i < MAXIMUM_BLOCK_SIZES
+		&& allocation_error_code; i++)	{
+		uint32_t blck_size =
+			1 << heap->physical_blocks_allocators[i]->blk_sz_shift;
+		/* Try span alloc on first available mem_block for non span check if
+		 * block size is sufficient.
+		 */
+		if (heap->span)	{
+			/* TO BE CONSIDERED!!!!!!!!!!!!!!!!!
```

Comment:
They probably should - I would mark it as a ```TODO:``` as this is searchable and shows up in some IDEs alongside our other TODOs

Reply:
This is not a to do , this is a rather blunt request for comment :D I wanted to mark it in code so rewievers would share opinion on that option/consideration. It will not end up in final code as it was still a draft pr.